starknet_py/net/contract/contract_function.py

```diff
@@ -0,0 +1,481 @@
+            calldata=calldata,
+            selector=self.get_selector(self.name),
+            _client=self.client,
+            _payload_transformer=self._payload_transformer,
+        )
+
+    async def call(
+        self,
+        *args,
+        block_hash: Optional[str] = None,
+        block_number: Optional[Union[int, Tag]] = None,
+        **kwargs,
+    ) -> TupleDataclass:
+        """
+        Call contract's function. ``*args`` and ``**kwargs`` are translated into Cairo calldata.
+        The result is translated from Cairo data to python values.
+        Equivalent of ``.prepare_call(*args, **kwargs).call()``.
+
+        :param block_hash: Block hash to perform the call to the contract at specific point of time.
+        :param block_number: Block number to perform the call to the contract at specific point of time.
+        :return: TupleDataclass representing call result.
+        """
+        return await self.prepare_call(*args, **kwargs).call(
+            block_hash=block_hash, block_number=block_number
+        )
+
+    def prepare_invoke_v1(
+        self,
+        *args,
+        max_fee: Optional[int] = None,
+        **kwargs,
+    ) -> PreparedFunctionInvokeV1:
+        """
+        ``*args`` and ``**kwargs`` are translated into Cairo calldata.
+        Creates a ``PreparedFunctionInvokeV1`` instance which exposes calldata for every argument
+        and adds more arguments when calling methods.
+
+        :param max_fee: Max amount of Wei to be paid when executing transaction.
+        :return: PreparedFunctionCall.
+        """
+
+        calldata = self._payload_transformer.serialize(*args, **kwargs)
+        return PreparedFunctionInvokeV1(
+            to_addr=self.contract_data.address,
+            calldata=calldata,
+            selector=self.get_selector(self.name),
+            max_fee=max_fee,
+            _contract_data=self.contract_data,
+            _client=self.client,
+            _account=self.account,
+            _payload_transformer=self._payload_transformer,
+        )
+
+    async def invoke_v1(
+        self,
+        *args,
+        max_fee: Optional[int] = None,
+        auto_estimate: bool = False,
+        nonce: Optional[int] = None,
+        **kwargs,
+    ) -> InvokeResult:
+        """
+        Invoke contract's function. ``*args`` and ``**kwargs`` are translated into Cairo calldata.
+        Equivalent of ``.prepare_invoke_v1(*args, **kwargs).invoke()``.
+
+        :param max_fee: Max amount of Wei to be paid when executing transaction.
+        :param auto_estimate: Use automatic fee estimation, not recommend as it may lead to high costs.
+        :param nonce: Nonce of the transaction.
+        :return: InvokeResult.
+        """
+        prepared_invoke = self.prepare_invoke_v1(*args, **kwargs)
+        return await prepared_invoke.invoke(
+            max_fee=max_fee, nonce=nonce, auto_estimate=auto_estimate
+        )
+
+    def prepare_invoke_v3(
```

Comment:
I'm aware of the fact that we agreed to use postfixes. I do wonder, however, **if in this particular case**, where (prepare)_invoke(v1 | v3) can be differed by max_fee/resource_bounds and this is a helper class, would it make sense to have one prepare_invoke and one invoke, that takes those values as Union and based on type it prepares/makes v1 or v3 tx? wdyt? I know it may look like breaking existing approach to interface, but I do wonder if it does make sense in a helper class

Reply:
I definitely don't like the current interface. However, setting a fee as `Union` may not be the best approach:
- A unified name for `max_fee` and `l1_resource_bounds` would be required, let's say just `fee`. While this could make sense for a user with plenty of context, it might be confusing in general
- Thinking long term we will need to incorporate other V3 fields such as `l2_resource_bounds`, `tip` etc. As a result, we would need to alter this interface again

One potential solution is to create `InvokeV1Params` and `InvokeV3Params` and then use 
```
prepare_invoke(params: InvokeParams)
```
However, this approach will break consistency with other parts of our codebase.

Reply:
Sure, it was just a rough idea either way. We can leave this as is.